Add Agentic Coding Tile (#25)
Ada-Developers-Academy/independent-ai-topics: agentic-coding/agentic-setup-steps.md

            *Fig. Navigation to the API Keys view in OpenRouter ([Full Size Image](assets/set-up/openrouter_workspaces.png))*
        
        2. On the "API Keys" page, click the "New Key" button towards the top right.
            ![OpenRouter's API Keys page with the "New Key" button highlighted](assets/set-up/openrouter_new_key_button.png)
            *Fig. Where to create a new key on OpenRouter's "API Keys" page ([Full Size Image](assets/set-up/openrouter_new_key_button.png))*
        
        3. In the pop up that displays, you are required to choose a name to help you distinguish between your keys. Nothing else is required to create a key, but we recommend looking through the options, since this is where we could add restrictions like a spending limit to the API key. 

            ![OpenRouter's UI for creating a new API key](assets/set-up/openrouter_new_key_ui.png)   
            *Fig. OpenRouter's form for creating a new API key*

        4. Once the key is created, it will only be displayed once. Copy the value and hold onto it securely since we'll need it in the next section!

### OpenRouter in VS Code

> To add OpenRouter to VS Code so we can choose models they provide from inside our IDE, we need to add OpenRouter as a provider and supply our API key.

Reviewer: The meaning of this sentence didn't land on the first read through for me. How about tweaking the wording a little bit like this: 
```suggestion
We need to add OpenRouter to VS Code so we can choose models they provide from inside our IDE. To do this, we need to add OpenRouter as a provider and supply our API key.
```

    ![VS Code UI for creating the group name for a set of models](assets/set-up/vscode_name_provider_group.png)   
    *Fig. VS Code's text field for naming the model group ([Full Size Image](assets/set-up/vscode_name_provider_group.png))*

5. A new prompt will appear at the top of the screen asking for your API key. Paste in your OpenRouter API key and press enter.
    ![VS Code's UI for adding a model provider's API key](assets/set-up/vscode_add_api_key.png)   
    *Fig. VS Code's text field for adding a provider's API key ([Full Size Image](assets/set-up/vscode_add_api_key.png))*

6. After a moment, the model settings ui should refresh and we should be able to see a new section and many more models that we now have access to. 
    - If at any point we need to update our API key or want to remove OpenRouter from VS Code, we can press the settings icon on the group name for options. 

    ![VS Code's model settings screen updated to show OpenRouter and its available models](assets/set-up/vscode_new_models_showing.png)   

>     *Fig. Updated screen showing OpenRouter models and settings to update or remove the model provider ([Full Size Image](assets/set-up/vscode_new_models_showing.png))*

Reviewer: I like that the alt text calls out this screenshot is from VS Code.

Could we update the caption to show the same for clarity?
```suggestion
    *Fig. VS Code's model settings screen showing OpenRouter models and settings to update or remove the model provider ([Full Size Image](assets/set-up/vscode_new_models_showing.png))*
```

### Choosing Models with OpenRouter

We now have access to multiple models and model providers! To see what models are curently free on OpenRouter, check out their ["Models" page](https://openrouter.ai/models) and use the filter options to select "Text" input and "Free" pricing. 
    ![OpenRouter's "models" page with filters for input type and cost highlighted](assets/set-up/openrouter_models_page_filtered.png)   
    *Fig. OpenRouter's "models" page with filters for input type and cost highlighted ([Full Size Image](assets/set-up/openrouter_models_page_filtered.png))*

When you see one that you like, we can search for that name under the model picker in VS Code!
    ![VS Code's model picker in the chat pane with the search bar highlighted](assets/set-up/vscode_model_picker_search_box.png)   
    *Fig. We can search models in VS Code using the search bar at the top of the model picker UI*

## Sandboxing and Scoping Access

> Before we dive into using the models we just made accessible, let's talk a little about security. We've established that LLMs and agents built on them operate on statistical text prediction, not by reasoning about consequences. If an agent's context makes a particular file write or command statistically likely, the agent will produce it, whether or not it's actually safe to run. 

Reviewer: How would you feel about being a little more verbose to spell out the idea being conveyed here? I'm not married to this exact wording, but it would be nice to provide some examples to folks that are unfamiliar with the topic.

```suggestion
Before we dive into using the models we just made accessible, let's talk a little about security. We've established that LLMs and agents built on them operate on statistical text prediction, not by reasoning about consequences. An agent's context might deem a particular file write statistically likely (such as editing an existing file, creating a new configuration file, or overwriting a file completely) even if it is not what is actually correct or secure. The agent might also produce a command that seems like a natural next step, but might not actually be safe to run.
```

Sandboxing is how we enforce constraints that prompts alone cannot reliably maintain. A **sandbox** is an isolated environment that constrains what an agent can reach. By scoping the access that AI agents have, we limit the potential blast radius of unexpected agent actions. 

### Why We Sandbox Agents

Without sandboxing, an agent with file system access can write to any location our user account has permission to write to. An agent with shell access can run any command available in that shell. This is rarely what we want! The most common problems sandboxes protect against are:

> - **Filesystem scope creep**: Without boundaries, an agent with file write access can write anywhere our user account has permission to write, including dotfiles, other project directories, or SSH keys.

Reviewer: Again, I think it could be nice to add a concrete example. I'm fine with a different example too if you think there's one that's a better fit.

```suggestion
- **Filesystem scope creep**: Without boundaries, an agent with file write access can write anywhere our user account has permission to write, including dotfiles, other project directories, or SSH keys.
    - Imagine asking an agent to clean up temporary files in your project. The agent interprets this broadly and starts deleting anything that looks temporary  like cache files in a sibling project directory that happened to be within reach. Without a sandbox boundary, 'clean up temporary files' turned into 'delete things outside the intended project'.
```

> - **Credential exposure**: Local environments often have API keys or cloud credentials sitting in environment variables or config files. An agent with unrestricted read access can pull these into its output or logs without any malicious intent behind it.

Reviewer: ```suggestion
- **Credential exposure**: Local environments often have API keys or cloud credentials sitting in environment variables or config files. An agent with unrestricted read access can pull these into its output or logs without any malicious intent behind it.
    - Imagine asking an agent to debug why an API call is failing and it prints out the full request, including a header with your API key, into a log file that later gets uploaded to a public GitHub repo for troubleshooting. The key is now public, and anyone can use it until it's revoked.
```

| **Container isolation** | Everything outside an explicitly mounted directory. The agent's filesystem, network, and process view are fully separated from the host machine. | Highest, typically requires Docker or a similar runtime tool |

**Application-level permissions** are the coarsest tool: they draw lines around categories of action, not specific paths, so "allow file writes" means the agent can write anywhere our account can. 

**OS-level isolation** tightens this by enforcing specific file path or system-call boundaries below the application, so even a misbehaving tool can't bypass the restriction. 

**Container isolation** is the strongest option, fully separating the agent's environment from the host, at the cost of needing Docker and some setup around what gets mounted into the container (what folder or volumes are visible inside the container).

For most development work, the choice comes down to risk profile and workflow needs:

| Level | Good for | Limitations |
|---|---|---|

> | Application permissions | Quick tasks, supervised sessions, low-sensitivity projects | Category-level only, below-application bypasses aren't blocked |

Reviewer: I think it could be helpful to elaborate on "below-application bypasses aren't blocked". 

```suggestion
| Application permissions | Quick tasks, supervised sessions, low-sensitivity projects | Category-level only, below-application bypasses aren't blocked meaning the permissions control which apps get access, but not what an agent does once inside. |
```

- We should be very careful about what commands we auto-approve when we're getting started. As examples, a command that creates or enables a `venv` is pretty safe, but allowing new dependencies to be installed without your awareness could both bloat the project and bring in packages with security vulnerabilities.

**Agent sandboxing** is VS Code's OS-level option. It's the setting `chat.agent.sandbox.enabled` on macOS and Linux, or `chat.agent.sandbox.enabledWindows` on Windows. We can find it by searching "chat.agent.sandbox.enabled" inside VS Code's settings: 

![VS Code's settings UI open showing the checkbox to turn on agent sandboxing](assets/set-up/vscode_enable_sandbox.png)
*Fig. The Sandbox setting enabled in VS Code's UI ([Full Size Image](assets/set-up/vscode_enable_sandbox.png))*

> We recommend that folks turn this setting on to give some security and start getting some exposure to how we can scope access for our tools. With it enabled, terminal commands run inside a boundary enforced by the operating system:

Reviewer: Could be nice to add a reminder about how to get to settings:

```suggestion
We recommend that folks turn this setting on to give some security and start getting some exposure to how we can scope access for our tools. You can access your settings with the shortcut `cmd` + `,`. With it enabled, terminal commands run inside a boundary enforced by the operating system:
```

- **File writes** are limited to the current working directory and its subdirectories.
- **Network access** is blocked by default, though we can allow specific domains if a command needs to reach one.

Because the boundary is enforced by the OS, commands that stay inside it run without a confirmation prompt. Anything that needs more than the sandbox allows surfaces a prompt asking whether to run it outside the sandbox instead. The isolation mechanism underneath differs by platform, but VS Code manages those details once the setting is turned on.

For our first agentic coding sessions, a reasonable combination is: 
1. keep the permission level on Default Approvals
2. turn on agent sandboxing

This pairs a human checkpoint before risky actions with an OS-enforced boundary in case something proceeds that we didn't anticipate, at the cost of a single settings change. As we mentioned earlier in the lesson, over time our tasks may grow more autonomous or start running unattended for longer stretches. At that point container isolation becomes worth the added setup, but VS Code's built-in sandboxing is enough to get us started.

## Next Steps & Notes

> Before moving on to the next lesson, we recommend opening up a project that you are familiar with in VS Code. Our goal is for students to keep a coding project up while reading through the lessons and pause to try things out as they go.

Reviewer: Would it be worth suggesting a couple projects or do you want to keep this pretty open ended?

```suggestion
Before moving on to the next lesson, we recommend opening up a project that you are familiar with (any project from Unit 1 would work well) in VS Code. Our goal is for students to keep a coding project up while reading through the lessons and pause to try things out as they go.
```

Author: I'm inclined to keep it more open ended, my hope being that folks will try these lessons out with different projects and have more to talk about/contrast when discussing how things went, but I like adding the Unit 1 projects note to help guide folks if they feel overwhelmed without some direction.

# Setting up Our Workspace

We'll be using two tools to provide free access to large language models for coding agents.
- [Github Copilot](https://github.com/features/copilot), through the GitHub Student Developer Pack
- [OpenRouter](https://openrouter.ai/)

> Once our accounts are set up, both of these will be used from inside VS Code, keeping our tools in one location so we don't need to leave our IDE to add agents to our workflow. We will also explore sandboxing as a way to secure agents, and look at what sandboxing controls VS Code provides.

Author: ```suggestion
Once our accounts are set up, both of these will be used from inside VS Code, keeping our tools in one location so we don't need to leave our IDE to add agents to our workflow. We will also explore sandboxing as a way to secure agents, and look at what sandboxing controls VS Code provides.

### !callout-info

## Prefer Copilot over OpenRouter

Copilot integrates better with VS Code and is what we recommend folks use first when trying out agentic coding. Credits are limited for Copilot access through the GitHub Student Developer Pack, so we have included OpenRouter set up as a fallback option in case we run out of Copilot credits. 

<br>

OpenRouter has rate limiting for free tiers that can cause issues with VS Code when those limits are hit and the agent cannot finish a request in progress. Typically we can recover by rolling back to the most recent checkpoint if we run into those kinds of problems.

### !end-callout
```

Sandboxing is how we enforce constraints that prompts alone cannot reliably maintain. A **sandbox** is an isolated environment that constrains what an agent can reach. By scoping the access that AI agents have, we limit the potential blast radius of unexpected agent actions. 

### Why We Sandbox Agents

Without sandboxing, an agent with file system access can write to any location our user account has permission to write to. An agent with shell access can run any command available in that shell. This is rarely what we want! The most common problems sandboxes protect against are:

- **Filesystem scope creep**: Without boundaries, an agent with file write access can write anywhere our user account has permission to write, including dotfiles, other project directories, or SSH keys.

>     - Imagine asking an agent to clean up temporary files in your project. The agent interprets this broadly and starts deleting anything that looks temporary  like cache files in a sibling project directory that happened to be within reach. Without a sandbox boundary, 'clean up temporary files' turned into 'delete things outside the intended project'.

Reviewer: Whoops, I included an extra space when I suggested this change

```suggestion
    - Imagine asking an agent to clean up temporary files in your project. The agent interprets this broadly and starts deleting anything that looks temporary like cache files in a sibling project directory that happened to be within reach. Without a sandbox boundary, 'clean up temporary files' turned into 'delete things outside the intended project'.
```
